### DISCUSSION TITLE
Servo Issues

### DISCUSSION THREAD

**Discussion by EAGrahamJr** [mattjlewis/diozero#114] [Q&A] [answered]
Attempting to use the `ServoDevice` on GPIO 18 and I'm not having a lot of luck. If I set up the `TOWERPRO_SG90` trim, the servo basically doesn't move at all. If I set up the _same_ pin with `PwmOutputDevice`, I can get the servo to move (using `setValue`), but has basically two settings:

```java
PwmOutputDevice device = new PwmOutputDevice(18);
device.setValue(.049f);  // swings to ~ 30 degrees
device.setValue(.05f);   // swings to ~ 170 degrees
```

This, on the other hand, does not engage the servo at all:
```java
ServoDevice device = ServoDevice.Builder(18).setTrim(TOWERPRO_SG90).build();
device.setAngle(30f);
```

Am I missing something or is this really more an artifact of the Software PWM implementation? (I _really_ don't like Python and was hoping `diozero` was my "Get Out of Jail" card.)

**Comment by mattjlewis** [comment]
I'll look into this ASAP

**Reply by EAGrahamJr** [reply]
Oh, there's no hurry! Ah, forgot - this is on a Raspberry Pi 3B+

On the _other hand_, I did find the very nice `PCA9685` device already implemented, so it's probably not that big a deal. :smirk:

Thanks for a very sane (and comprehensive) library!

**Comment by mattjlewis** [comment]
FYI I've just pushed an update that includes a few fixes in this area. There's a new set of unit tests for servo devices via native servo control (e.g. Picon Zero, Arduino), as well as hardware and software PWM.
Are you able to pull and do a local build? I _might_ be able to test with real servo devices this week, otherwise it won't be until after Xmas. I made some notes [here](https://www.diozero.com/internals/rpinotes.html#hardware-pwm) for enabling hardware PWM on the Raspberry Pi - this is detected automatically and software PWM is used as a last resort.

**Reply by EAGrahamJr** [reply]
Yes, I can build locally :smiley: - I can loop back around and try out the servo this week. The notes on the GPIO pins are **really** helpful.

**Comment by EAGrahamJr** [comment, marked as answer]
Servo fix verified for **both** hardware (18) and software (23) PWM. Both produce full sweeps from 0 to "real close to" 180 degrees.

**Reply by mattjlewis** [reply]
That's brilliant, thank you for testing. How jittery was the software PWM controlled servo?

**Reply by EAGrahamJr** [reply]
They both had moments of jitter (which was oddly not reproducible unless I ran the **exact** same sequence), but both were overall pretty steady. I was using the `angle` setting, so I would imagine that rounding errors probably contributed somewhat.
